File: algorithm/basic/가까운수.js
const array = [3,10,28]
const n = 20


function solution(array, n){
    
    const abs_array = array.map(el => Math.abs(el-n))
    const filtered = array.filter(el => Math.abs(el-n)===Math.min(...abs_array)) // |array-n|이 가장 작은 array의 요소만 필터링.
    
    return  Math.min(...filtered) // 그 중에서도 가장 작은 수를 고름.
}

console.log(solution(array,n))

// 생각
// 각 원소 - n 을 한 array. 중에서 가장 작은 수를 찾고, 이를 index로 원 array에서 원소를 찾는다.
// 가장 가까운 수가 여러 개일 경우 더 작은 수를 return 합니다. > 이 조건에서 걸림. 
// 1) 절대값이 가장 작고, 그 값이 같은 수 중 2) 가장 작은 수만 return

File: algorithm/basic/가까운수.py
array = [3, 10, 28]
n = 20

solution=lambda a,n:sorted(a,key=lambda x:(abs(x-n),x))[0]


print(solution(array,n))

# abs(x-n) > 3,10,28 에서 각각 20을 뺀 절대값
# 그리고 동일한 차이를 가진다면 원래 값 x를 기준으로 
# sorted (정렬한 것)를 list로 반환
# 마지막으로 [0] 가장 첫번째, 가장 작은 원소를 반환.